Store Orders: Enable saving of new orders

ryelle opened the pull request:
When saving a new order, the order needs to be sent to a generic `/orders` endpoint (rather than the `orders/:id` endpoint for updating). This PR updates `sendOrder` to be context-aware, if the order saved has a placeholder ID, it's sent to the ID-less creation endpoint.

Two other changes in this PR:

- Enable UPDATE actions to save placeholder orders, so that we can check the isSaving status of new orders
- Switch to passing onSuccess and onFailure callbacks through the `saveOrder` function, so that we can trigger different behavior after saving a new order vs updating an existing one – the major difference being new orders need a page redirect, existing orders don't.

**To test**

- Visit `/store/order/:site` to create an order
- Fill out some info - nothing is required yet, I think, so fill in as much or as little as you want
- Save your order
- Should redirect to the single view of your just-created order, with a success message
- Trigger an error state by deactivating wc-api-dev (or interfering with the API in some other way)
- Try to save, you should see an error notice and your changes should stay on screen in the pending edit state

Make sure the tests still pass:

`npm run test-client client/extensions/woocommerce`

See #15679

timmyc commented on client/extensions/woocommerce/state/sites/orders/reducer.js:
```diff
@@ -105,7 +106,7 @@ export function items( state = {}, action ) {
 		case WOOCOMMERCE_ORDER_REQUEST_SUCCESS:
 		case WOOCOMMERCE_ORDER_UPDATE_SUCCESS:
-			orders = { [ action.orderId ]: action.order };
+			orders = { [ action.order.id ]: action.order };
```
Any chance we could have a null `action.order` ?

timmyc replied:
And do we ever clear out the placeholder order status?

ryelle replied:
No, it shouldn't be, because [we check for order.id](https://github.com/Automattic/wp-calypso/blob/master/client/extensions/woocommerce/state/sites/orders/actions.js#L115) before firing `WOOCOMMERCE_ORDER_UPDATE_SUCCESS`.

We also clear the edit state (and thereby the placeholder) for a site when a successful save happens. We don't technically clear the isUpdating state for anything, even editing existing orders – once an edit is saved, it's set to false and just lives in the state. We could revisit this in a future PR if you think this might lead to performance issues?

timmyc commented on client/extensions/woocommerce/app/order/create.js:
```diff
@@ -50,8 +52,18 @@ class Order extends Component {
-		const { siteId, order } = this.props;
-		this.props.saveOrder( siteId, order );
+		const { site, siteId, order, translate } = this.props;
+		const onSuccess = ( dispatch, orderId ) => {
```
Should the messaging here be contextual. Like when creating a new order, should the messaging say "Order created." / "Unable to create order."? I think the current verbiage works in both cases, but figured it would be worth asking :)

budzanowski commented on client/extensions/woocommerce/state/sites/orders/reducer.js:
```diff
@@ -78,12 +78,13 @@ export function isQueryLoading( state = {}, action ) {
  * @return {Object}        Updated state
  */
 export function isUpdating( state = {}, action ) {
+	const orderId = isNumber( action.orderId ) ? action.orderId : JSON.stringify( action.orderId );
```
Maybe use `isFinite` instead off `isNumber` for security against `Infinity`, `-Infinity`, and `NaN`.

ryelle replied:
Thanks for the tip - I doubt we'll need to worry about `Infinity`, but something that catches `NaN` sounds useful 🙂

ryelle commented:
@timmyc I've updated the PR with your feedback & fixed the tests - I changed to the `onSuccess`/`onFailure` method without updating tests, so they've been updated in 1cfd712.

ryelle commented:
> Should I create an issue for the "Pay Via " empty bug?

I've added a note about it in a new issue I added for the payment process: #20794
